http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java

```diff
@@ -1491,181 +1454,360 @@ private <I> Publisher<URI> resolveURI(io.micronaut.http.HttpRequest<I> request,
+                // local, and is more likely to be wrong than any reactive context we are fed.
+                if (c.hasKey(ServerRequestContext.KEY)) {
+                    return c;
+                } else {
+                    return c.put(ServerRequestContext.KEY, parentRequest);
+                }
+            });
         }
+        return responseMono;
+    }
 
-        if (log.isTraceEnabled()) {
-            traceRequest(request, nettyRequest);
+    private Mono<HttpResponse<?>> sendRequestWithRedirectsNoFilter(
+        io.micronaut.http.HttpRequest<?> parentRequest,
+        @Nullable BlockHint blockHint,
+        MutableHttpRequest<?> request,
+        BiFunction<MutableHttpRequest<?>, NettyClientByteBodyResponse, ? extends Mono<? extends HttpResponse<?>>> readResponse
+    ) {
+        RequestKey requestKey;
+        try {
+            requestKey = new RequestKey(this, request.getUri());
+        } catch (Exception e) {
+            return Mono.error(e);
         }
+        // first: connect
+        return connectionManager.connect(requestKey, blockHint)
+            .flatMap(poolHandle -> {
+                // build the raw request
+                request.setAttribute(NettyClientHttpRequest.CHANNEL, poolHandle.channel);
 
-        new ByteBodyRequestWriter(nettyRequest, byteBody).write(poolHandle);
-        responsePromise.addListener((Future<MutableHttpResponse<?>> future) -> {
-            if (future.isSuccess()) {
-                emitter.next(future.getNow());
-                emitter.complete();
-            } else {
-                emitter.error(future.cause());
-            }
-        });
-    }
+                URI requestURI = request.getUri();
+                boolean permitsBody = io.micronaut.http.HttpMethod.permitsRequestBody(request.getMethod());
+                NettyByteBody byteBody;
+                try {
+                    byteBody = buildNettyRequest(
+                        request,
+                        requestURI,
+                        request
+                            .getContentType()
+                            .orElse(MediaType.APPLICATION_JSON_TYPE),
+                        permitsBody,
+                        poolHandle.channel.eventLoop()
+                    );
+                } catch (HttpPostRequestEncoder.ErrorDataEncoderException e) {
+                    poolHandle.release();
+                    return Mono.error(e);
+                }
 
-    private ByteBuf charSequenceToByteBuf(CharSequence bodyValue, MediaType requestContentType) {
-        return byteBufferFactory.copiedBuffer(
-                bodyValue.toString().getBytes(
-                        requestContentType.getCharset().orElse(defaultCharset)
-                )
-        ).asNativeBuffer();
-    }
+                // send the raw request
+                return sendRawRequest(poolHandle, request, byteBody, c -> handleResponseError(request, c));
```

Comment:
More correct would be to flatMap response of this method instead of the outer one to avoid an extra call for the error path

Reply:
it's functionally identical